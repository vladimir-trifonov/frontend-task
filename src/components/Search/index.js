import React, { useState, useEffect, useRef } from 'react'
import { Search, Icon } from 'semantic-ui-react'
import debouce from 'lodash.debounce'

export default ({ onSearch }) => {
  const [search, setSearch] = useState('')
  const ref = useRef()

  useEffect(() => {
    ref.current = debouce(onSearch, 300)
  }, [onSearch])

  const handleSearch = (q) => {
    setSearch(q)
    ref.current(q)
  }

  const handleOnClick = () => handleSearch('')
  const handleOnSearchChange = (e, { value }) => handleSearch(value)

  return <Search
    value={search}
    {...search && { icon: <Icon name='delete' link onClick={handleOnClick} /> }}
    open={false}
    onSearchChange={handleOnSearchChange}
  />
}
