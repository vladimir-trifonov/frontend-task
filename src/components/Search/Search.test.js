/* global describe, jest, test, expect, beforeAll, afterAll */
/* eslint-disable import/first */
import React from 'react'
import Search from './'
import { mount } from 'enzyme'
import { create } from 'react-test-renderer'

jest.mock('lodash.debounce')
import debouce from 'lodash.debounce'

describe('Search component', () => {
  test('it matches the snapshot', () => {
    const component = create(<Search onSearch={() => {}} />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('Functional testing', () => {
    beforeAll(() => {
      jest.mock('lodash.debounce')
      debouce.mockImplementation(fn => fn)
    })

    afterAll(() => {
      jest.unmock('lodash.debounce')
    })

    test('should invoke the onSearch callback', () => {
      const mockFn = jest.fn()
      const search = mount(<Search onSearch={mockFn} />)

      // manually enters the search text
      search.find('input').simulate('change', { target: { value: 'abc' } })
      expect(mockFn.mock.calls[0][0]).toBe('abc')
    })
  })
})
