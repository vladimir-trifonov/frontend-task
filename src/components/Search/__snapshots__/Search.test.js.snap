// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component it matches the snapshot 1`] = `
<div
  className="ui search"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="ui icon input"
  >
    <input
      autoComplete="off"
      className="prompt"
      onChange={[Function]}
      onClick={[Function]}
      tabIndex="0"
      type="text"
      value=""
    />
    <i
      aria-hidden="true"
      className="search icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="results transition"
  >
    <div
      className="message empty"
    >
      <div
        className="header"
      >
        No results found.
      </div>
    </div>
  </div>
</div>
`;
